backend/src/main/java/com/google/sps/TripCRUD.java

```diff
@@ -0,0 +1,138 @@
+    datastore.put(tripEntity);
+    UserCrud.addTripId(email, tripEntity.getKey().getId());
+    return tripEntity;
+  }
+
+  /**
+   * Converts string tripData to Trip Entity
+   *
+   * @param tripData tripData string representation of a trip json
+   * @param tripId trip id
+   * @param tripKey trip key, can be null
+   * @return tripEntity from tripData
+   */
+  public static Entity toEntity(String tripData, String tripId, Key tripKey) {
+    Entity tripEntity = tripId == "" ? new Entity("Trip") : new Entity("Trip", tripId, tripKey);
+    setProperties(tripEntity, tripData);
+    return tripEntity;
+  }
+
+  /**
+   * Mutates trip entity properties given string representation of tripData json
+   *
+   * @param tripEntity Trip Entity
+   * @param tripData tripData string representation of a trip json
+   */
+  private static void setProperties(Entity tripEntity, String tripData) {
+    JsonParser parser = new JsonParser();
+    JsonElement jsonElement = parser.parse(tripData);
+    JsonObject jsonObject = jsonElement.getAsJsonObject();
+
+    tripEntity.setProperty(
+        "isOptimized", Boolean.parseBoolean(jsonObject.get("isOptimized").toString()));
+    tripEntity.setProperty("searchText", jsonObject.get("searchText").toString());
+    tripEntity.setProperty("tripName", jsonObject.get("tripName").toString());
+    ArrayList<EmbeddedEntity> attractions = new ArrayList<EmbeddedEntity>();
+
+    for (JsonElement attractionElement : jsonObject.getAsJsonArray("attractions")) {
+      JsonObject attraction = attractionElement.getAsJsonObject();
+      EmbeddedEntity embeddedAttraction = new EmbeddedEntity();
+      embeddedAttraction.setProperty("attractionName", attraction.get("attractionName").toString());
+      embeddedAttraction.setProperty("photoReference", attraction.get("photoReference").toString());
+      embeddedAttraction.setProperty(
+          "routeIndex", Integer.parseInt(attraction.get("routeIndex").toString()));
+      embeddedAttraction.setProperty("coordinates", attraction.get("coordinates").toString());
+      attractions.add(embeddedAttraction);
+    }
+    tripEntity.setProperty("attractions", attractions);
+  }
+
+  /**
+   * Finds a trip entity by id
+   *
+   * @param tripId id of the trip to find
+   * @throws EntityNotFoundException if trip entity is not found
+   * @return Trip entity
+   */
+  public static Entity readTrip(String tripId) throws EntityNotFoundException {
+    Key entityKey = KeyFactory.createKey("Trip", Long.parseLong(tripId));
+    return datastore.get(entityKey);
+  }
+
+  /**
+   * Converts Trip Entity to Json
+   *
+   * @param tripEntity Trip Entity
+   * @return JsonObject of trip
+   */
+  public static JsonObject toJson(Entity tripEntity) {
+    JsonObject jsonTrip = new JsonObject();
+    jsonTrip.addProperty("tripId", Long.toString(tripEntity.getKey().getId()));
+    jsonTrip.addProperty(
+        "isOptimized", Boolean.parseBoolean(tripEntity.getProperty("isOptimized").toString()));
```

Comment:
Entity.getProperty returns an object which you can cast to the correct type, so here (and below) can just be
```suggestion
    jsonTrip.addProperty("isOptimized", (Boolean) tripEntity.getProperty("isOptimized"));
```

Reply:
okidoki done